### DISCUSSION TITLE
Accessing material property quadrature values on element boundaries

### DISCUSSION THREAD

**Discussion by ttruster** [idaholab/moose#21951] [Q&A Modules: General] [answered]
Hello,

I've been working on programming some interface (Nitsche/DG) applications, and I noticed from modeling an analytical solution that I am not correctly requesting the values of properties located at the boundary quadrature points.
Instead, the values I am getting correspond to the centroid of the element and neighbor, duplicated along all _qp of the given interface kernel integration.

I have checked this in Mortar Constraints, DG Kernels, and Interface Material/Kernels, and in all cases, when I request the property according to:
```
    _heat_flux_e(getMaterialProperty<RealVectorValue>(_base_name_e + "grad_t")),
    _heat_flux_n(getNeighborMaterialProperty<RealVectorValue>(_base_name_n + "grad_t")),
```
this leads to the element centroid value (of course different for the two elements when the analytical solution and the numerical solution have a non-constant gradient). Here, the material property is a simple test of just copying the value of the temperature field gradient into a material property, declared as:

```
  : DerivativeMaterialInterface<Material>(parameters),
    _temp(coupledValue("temp")),
    _grad_temp(coupledGradient("temp")),
    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
    _grad_t(declareProperty<RealVectorValue>(_base_name + "grad_t")),
```

I've read here at "Evaluation of Material Properties on Element Faces" from https://mooseframework.inl.gov/syntax/Materials/index.html that there are supposedly values computed for _face and _neighbor. And at one point when I had an error in my input files, the error codes mentioned all 3 of the material variations.

However, looking up the material getMaterialProperty<RealVectorValue>(_base_name_e + "grad_t_face") produced an error message.

Can you mention the proper way to request the boundary values of the material?

I can provide some simple source code (penalty method) which when run in debug with breakpoints shows that the values of the material property gradient do not match the temperature variable gradient at the interface point.

Thanks,

Tim

**Comment by jiangwen84** [comment]
Hi Tim, I have been using DG and Interface Kernel to access material properties. I do not see any issues with that. What you did seems to be correct to me. It should produce the correct material properties on the boundary/interface. If you could provide me your code and simple test, I can take a quick look.

**Reply by ttruster** [reply]
Hi Wen,
I pulled out the files for the three sets of objects I've developed, and included a the test file for each and for CG-Kernel alone to show the analytical result. When the flag 'use_flux_penalty' is false, the dgkernel produces the exact solution, but for true then the gradient penalty term causes the solution to deviate.
Let me know your thoughts, thanks,
Tim
[beaver3.tar.gz](https://github.com/idaholab/moose/files/9446256/beaver3.tar.gz)

**Comment by jiangwen84** [comment, marked as answer]
@ttruster  I do not see any issues with your code. I printed out some values inside ```TestDGBase::precalculateMaterial()```. The material properties you asked for have the same values as the temperature variable gradient. Do I misunderstand your question? 
```
_heat_flux_e[_qp] = (x,y,z)=(     0.5,     -0.5,        0)
_heat_flux_n[_qp] = (x,y,z)=(     1.5,     -0.5,        0)
_grad_u[_qp] = (x,y,z)=(     0.5,     -0.5,        0)
_grad_u_neighbor[_qp] = (x,y,z)=(     1.5,     -0.5,        0)  
```

**Reply by ttruster** [reply]
@jiangwen84 Yeah, I now checked also in debug and see the same value for _heat_flux and _grad_t.

Sorry for my lengthy response below...

I guess the issue is, it seems like the values of the variables at the physical quadrature point location don't seem to match with the analytical solution on the iteration after the first residual, and the analysis converges in one iteration which says that the residual and Jacobian are consistent with each other.
For the value use_flux_penalty = false, then the computed solution matches nodally with the analytical solution, as seen in Paraview from a PlotOverLine. But with use_flux_penalty = true, the computed result is different.

The term in the weak form is [(grad_test_e - grad_test_n) * normal] * penalty * [(grad_u_e - grad_u_n) * normal].
For the analytical solution ( u=x^2-y^2 and du = (2*x, -2*y) ), it is continuous everywhere and the gradient is continuous, so the jump should be zero at the converged solution.

Yet in both your printout and my tests, the values of the gradient on the e and n sides are not equal to each other.

That's what I'm not understanding: why does the reported value of the grad_u and grad_u_neighbor variable not seem to correspond with the boundary quadrature point location, even when the simulation (when converged and viewed in Paraview) has continuous values?
This apparent inconsistent location value is the explanation I am seeing for why the use_flux_penalty = true/false give different results.

Here is an example for a 2x1 element model:
_heat_flux_n[_qp]
_coords
[0]:1
[1]:0
[2]:0
_heat_flux_e[_qp]
_coords
[0]:-1
[1]:0
[2]:0
_grad_u_neighbor[_qp]
_coords
[0]:1
[1]:0
[2]:0
_grad_u[_qp]
_coords
[0]:-1
[1]:0
[2]:0
_q_point[_qp]
_coords
[0]:0
[1]:-0.28867513459481287
[2]:0

Technically, the du/dx value should be 0.0 on both sides, but it's not...

**Reply by jiangwen84** [reply]
I found one issue with your DG input file. To use DG, you need to use a discontinuous variable. For that, you can use 
```
[Variables]
  [./u]
    order = FIRST
    family = MONOMIAL
  [../]
[]
```
<img width="1111" alt="Untitled" src="https://user-images.githubusercontent.com/11078681/187316586-adde6e41-15c9-4ea5-a815-e172341d4807.png">

This is the output in paraview. Because the variable is first order, the gradient is constants. Looks like the grad_u and grad_u_neighbor are not the same and the output from the code seems to be correct. What do you think?

**Reply by ttruster** [reply]
Yep, that's what late night work does to you...

Found the problem: I had wanted to use SECOND order shape functions for all the tests, since the solution is quadratic. I forgot that you need both QUAD9 (in all the files) AND a SECOND variable type (in none of the files).
I also forgot (or didn't know; thought it was only 1d) that the continuous Galerkin method for heat equation will also interpolate the solution in simple cases like this.

So: with SECOND variables for all the tests, I indeed recover the continuous field and gradient field like I was expecting.

FYI: the DG solution "runs" fine with either biquadratic LAGRANGE or quadratic MONOMIAL functions, as long as I switch the BC type from FunctionDirichletBC to DGFunctionDiffusionDirichletBC. Not sure if that is intended or not; I run DG problems all the time in my other FE codes with Lagrange functions.

So: Thanks for the objective set of eyes! Always helps for finding these little things.

**Reply by jiangwen84** [reply]
@ttruster  In MOOSE if you want to use DG, the family type has to be either `L2_LAGRANGE` or `MONOMIAL`. If you use `LAGRANGE`, you will have a continuous mesh and your DG kernel become useless. 

`DGFunctionDiffusionDirichletBC` weakly enforces the Dirichlet BC at the boundary. My understanding is for `L2_LAGRANGE` or `MONOMIAL` type of shape functions, the nodes are not placed on the boundaries, so you have to weakly enforce them instead of directly set their values.
